cunumeric/deferred.py

```diff
@@ -441,27 +688,63 @@ def set_item(self, key, rhs):
         assert self.dtype == rhs.dtype
         # Check to see if this is advanced indexing or not
         if is_advanced_indexing(key):
+            view_copy = False
             # Create the indexing array
-            index_array = self._create_indexing_array(key)
-            if index_array.shape != rhs.shape:
-                raise ValueError(
-                    "Advanced indexing array does not match source shape"
-                )
-            if self.ndim != index_array.ndim:
-                raise NotImplementedError(
-                    "need support for indirect partitioning"
+            copy_needed, store, index_array = self._create_indexing_array(
```

Comment:
The meaning of `copy_needed` seems to be different in this case vs the `get_item` case. On `get_item`, `copy_needed == False` seems to signify the special case where the operation was handled in full by the `advanced_indexing` task because we were dealing with a boolean index array. What does it mean in this case? It can't mean the same thing, because you always need to perform a shuffle after this point.

Reply:
I think we can avoid checking on the `copy_needed` in the `set_item` case. We get `copy_needed == False` only in the case when `AdvancedIndexing` task was executed and, therefore, when there was no transformation on the `self`. I believe we can just do:
```
             if self.base.transform.bottom:
                    lhs = self
                else:
                    # if  store is transformed we need to to return a copy of
                    # the store since Copy operation can't be done on
                    # the store with transformation
                    ... copy logic here
                    lhs = store_copy
                    view_copy = True
```

Reply:
Actually, don't you already handle this in `_create_indexing_array`?

Reply:
If I'm understanding this correctly, the `_create_indexing_array` call is taking care of creating any copies that are required for `lhs` at this point. If this is correct, then I think you should just return the final array from `_create_indexing_array`, not just the store (e.g. on line 532 you do `rhs, store = self.copy_store(store)`, I would return `rhs` instead of `store`). And at this point you should be able to use that array as `lhs`.

Reply:
Yes, thank you!